File: yh_star/scripts/yh_star_sub.py
#!/usr/bin/python3
# -*- coding: utf-8 -*-
import rospy
from yh_star.msg import YhStarMsg

def msgCallback(msg):
    n = msg.data
    # n/2 => 실수 + 정수 ===> 숫자형

    for i in range(1, n//2+1): # 목 연산자//, i = 1 ~ n//2까지 들어감 
        for j in range(i): # j = 0
            print("*", end='') # end에는 줄바꿈이 기본으로 있지만 엔터치지 말고 별을 찍으라고 ''을 입력시킴
        print()
    for i in range(n//2+1, n+1): # i = n//2부터 n까지 들어감 n에 숫자를 넣어서 비교를하고, 숫자를세는 기준은 처음부터센다. ex. n = 3 3//2+1, 3+1 = 2, 4 즉, 2, 3을 i에 집어넣고 차례대로 계산해보면됨.
        for j in range(n-i+1):
            print("*", end='')
        print()

def star_printer():
    rospy.init_node("yh_star_sub")
    rospy.Subscriber("yh_star_topic", YhStarMsg, msgCallback)
    rospy.spin()

if __name__ == "__main__":
    star_printer()

File: yh_star/src/yh_star_sub.cpp
#include "ros/ros.h"
#include "yh_star/YhStarMsg.h"

void msgCallback(const yh_star::YhStarMsg::ConstPtr& msg)
{
    int n = msg->data; 
    for(int i = 0; i < n/2; i++) //별찍기. n/2까지
    {
        for(int j = 0; j < i+1; j++) //첫줄에 무조건 별을찍기위해
        {
            printf("*"); //가로로 쭉찍음
        }
        printf("/n"); //수행하면 엔터치면서 줄바꿈
    }

    for(int i=n/2; i<n; i++) //별을 줄임
    {
        for(int j = n; j > i; j--) // 
        {
            printf("*");
        }
        printf("\n");
    }
}

// n/2 주어진 숫자의 절반만큼까지는 증가. 그 다음부터는 감소.
// * 3나누기 2는 1이다. (정수기때문에 소수점은 상관없다.)

int main (int argc, char** argv)
{
    ros::init(argc, argv, "yh_star_sub");
    ros::NodeHandle nh;

    ros::Subscriber sub = nh.subscribe("yh_star_pub", 10, msgCallback);

    ros::spin();

    return 0;
}